packages/spring/src/index.ts

```diff
@@ -0,0 +1,215 @@
+ */
+export type WidenSpringTarget<T> = T extends number ? number : T;
+
+export type SpringSetterOptions = { hard?: boolean; soft?: boolean | number }
+export type SpringSetter<T> = (
+  newValue: T | ((prev: T) => T),
+  opts?: SpringSetterOptions,
+) => Promise<void>;
+
+/**
+ * Creates a signal and a setter that uses spring physics when interpolating from
+ * one value to another. This means when the value changes, instead of
+ * transitioning at a steady rate, it "bounces" like a spring would,
+ * depending on the physics paramters provided. This adds a level of realism to
+ * the transitions and can enhance the user experience.
+ *
+ * `T` - The type of the signal. It works for the basic data types that can be
+ * interpolated: `number`, a `Date`, `Array<T>` or a nested object of T.
+ *
+ * @param initialValue The initial value of the signal.
+ * @param options Options to configure the physics of the spring.
+ * @returns Returns the spring value and a setter.
+ *
+ * @example
+ * const [progress, setProgress] = createSpring(0, { stiffness: 0.15, damping: 0.8 });
+ */
+export function createSpring<T extends SpringTarget>(
+  initialValue: T,
+  options: SpringOptions = {},
+): [Accessor<WidenSpringTarget<T>>, SpringSetter<WidenSpringTarget<T>>] {
+  const [signal, setSignal] = createSignal(initialValue);
+  const { stiffness = 0.15, damping = 0.8, precision = 0.01 } = options;
+
+  if (isServer) {
+    return [signal as any, ((param: any, opts: SpringSetterOptions = {}) => {
+      if (opts.hard || signal() == null || (stiffness >= 1 && damping >= 1)) {
+        setSignal(param);
+        return Promise.resolve();
+      }
+      return new Promise(() => {});
+    }) as any]
+  }
+
+  let value_current = initialValue;
+  let value_last = initialValue;
+  let value_target = initialValue;
+  let inv_mass = 1;
+  let inv_mass_recovery_rate = 0;
+  let raf_id = 0
+  let settled = true
+  let time_last = 0;
+  let time_delta = 0
+  let resolve = () => {}
+
+  const cleanup = onCleanup(() => {
+    cancelAnimationFrame(raf_id)
+    raf_id = 0
+    resolve()
+  })
+
+  const frame: FrameRequestCallback = time => {
+    time_delta = (time - time_last) * 60 / 1000
+    time_last = time
+
+    inv_mass = Math.min(inv_mass + inv_mass_recovery_rate, 1)
+    settled = true
+
+    let new_value = tick(value_last, value_current, value_target)
+    value_last = value_current
+    setSignal(value_current = new_value)
+
+    if (settled) {
+      cleanup()
+    } else {
+      raf_id = requestAnimationFrame(frame)
+    }
+  }
+
+  const set: SpringSetter<T> = (param, opts = {}) => {
+    value_target = typeof param === "function" ? param(value_current) : param;
+
+    if (opts.hard || (stiffness >= 1 && damping >= 1)) {
+      cleanup()
+      setSignal(_ => value_current = value_last = value_target);
+      return Promise.resolve();
+    }
+
+    if (opts.soft) {
+      inv_mass_recovery_rate = 1 / (typeof opts.soft === "number" ? opts.soft * 60 : 30);
```

Comment:
This hard-codes an assumed 60 fps into the inverse mass recovery feature. On different hardware, `opts.soft` will behave differently (inv_mass will return to 1 more quickly with higher framerate) and more importantly the `velocity` will be incorrect.

I'm not sure the best fix for fps-independence here, but in my own adaptation I have gone for tracking the fps as `Math.max(fps, 1000 / (time - timeLast))` in the `frame` function above

Reply:
Turns out tracking fps this way is very flawed. At best it gives an inaccurate value (consistently getting ~180 fps when I should really be getting 166) and at worst is can go to extreme values and be very inconsistent.

I can look into frame-independent impls unless somebody else wants to jump in with a solution

Reply:
My intuition would be to keep the recovery rate as a constant, only depending on `opts.soft`.
And use time delta in `inv_mass = Math.min(inv_mass + inv_mass_recovery_rate * time_delta, 1)`
But I'm curious about other directions.
I'll try to add a test for framerate-independence since we are mocking the raf anyway.

Reply:
I was able to fix the `inv_mass_recovery_rate` to actually recover in the correct amount of time, but couldn't accomplish the same for the `tick` function

The frame-dependent `time_delta` is baked into this implementation in `tick`. It doesn't quite cancel out even though `d` is multipled by `time_delta` in the end either, but I'm not totally sure what is supposed to be going on instead.

https://github.com/pqml/spring/blob/master/lib/Spring.js - this looks promising, it is based on picking a timestep, checking how many of those timesteps have passed since the last frame, and solving for the position that many times